src/incremental/__init__.py

```diff
@@ -360,36 +348,173 @@ def getVersionString(version):  # type: (Version) -> str
     """
-    Get the version from the package listed in the Distribution.
+    Load the current version from {path}/_version.py.
     """
-    if not value:
+    version_info = {}  # type: Dict[str, Version]
+
+    versionpath = os.path.join(path, "_version.py")
+    with open(versionpath, "r") as f:
+        exec(f.read(), version_info)
+
+    return version_info["__version__"]
+
+
+def _get_setuptools_version(dist):  # type: (_Distribution) -> None
+    """
+    Setuptools integration: load the version from the working directory
+
+    This function is registered as a setuptools.finalize_distribution_options
+    entry point [1]. It is a no-op unless there is a pyproject.toml containing
+    an empty [tool.incremental] section.
+
+    @param dist: An empty C{setuptools.Distribution} instance to mutate.
+
+    [1]: https://setuptools.pypa.io/en/latest/userguide/extension.html#customizing-distribution-options
+    """
+    config = _load_pyproject_toml("./pyproject.toml")
+    if not config:
         return
 
-    from distutils.command import build_py
+    dist.metadata.version = _existing_version(config.path).public()
+
+
+def _get_distutils_version(dist, keyword, value):  # type: (_Distribution, object, object) -> None
+    """
+    Distutils integration: get the version from the package listed in the Distribution.
+
+    This function is invoked when a C{setup.py} calls C{setup(use_incremental=True)}.
+
+    @see: https://setuptools.pypa.io/en/latest/userguide/extension.html#adding-arguments
+    """
+    if not value:  # use_incremental=False
+        return  # pragma: no cover
+
+    from setuptools.command import build_py  # type: ignore
 
     sp_command = build_py.build_py(dist)
     sp_command.finalize_options()
 
-    for item in sp_command.find_all_modules():  # type: ignore[attr-defined]
+    for item in sp_command.find_all_modules():
         if item[1] == "_version":
-            version_file = {}  # type: Dict[str, Version]
+            package_path = os.path.dirname(item[2])
+            dist.metadata.version = _existing_version(package_path).public()
+            return
 
-            with open(item[2]) as f:
-                exec(f.read(), version_file)
+    raise Exception("No _version.py found.")  # pragma: no cover
 
-            dist.metadata.version = version_file["__version__"].public()
-            return None
 
-    raise Exception("No _version.py found.")
+def _load_toml(f):  # type: (BinaryIO) -> Any
+    """
+    Read the content of a TOML file.
+    """
+    # This import is deferred to avoid a hard dependency on tomli
+    # when no pyproject.toml is present.
+    if sys.version_info > (3, 11):
+        import tomllib
+    else:
+        import tomli as tomllib
 
+    return tomllib.load(f)
 
-from ._version import __version__  # noqa: E402
+
+@dataclass(frozen=True)
+class _IncrementalConfig:
+    """
+    @ivar package: The package name, capitalized as in the package metadata.
```

Comment:
FYI, you can just put these docstrings inline underneath the attribute declaration annotation and pydoctor will pick 'em up